### PR TITLE
Add slug to document URLs

### PR DESCRIPTION
**Pull request by asaunier**
Fixes https://github.com/c2corg/v6_ui/issues/109
Fixes https://github.com/c2corg/v6_ui/issues/204

Document URL can be of 3 forms:
* {type}/{id}
* {type}/{id}/{lang}
* {type}/{id}/{lang}/{slug}

The standard one is the third one. When loaded, the 2 others should redirect to it. If the lang is missing, the "best" lang should be found out by the API.

As for now the PR use the standard ``Document.get_document()`` function to get the lang and title of the document (in order to figure the slug out). This is not optimal since this triggers a full document retrieval. We need to create a special service for that in the API. See https://github.com/c2corg/v6_api/issues/364

The serve-side generated slugs uses https://github.com/un33k/python-slugify (there's also https://github.com/dimka665/awesome-slugify both seem quite equivalent).

We also need to slugify the document URLs in the cards (client-side). I have found those 2 angular libs:
* https://github.com/paulsmith/angular-slugify
* https://github.com/lucasconstantino/angular-slug

TODO
* [x] server-side slugs
* [x] client-side slugs
* [x] fix conflicts with the caching system
* [x] update tests

### CODE REVIEW

**Comment by tsauerwein** [comment]
> We also need to slugify the document URLs in the cards (client-side). I have found those 2 angular libs:

What about generating the slugs in the API and including them for each document? This way we are sure the same slugs are used. If we use a different library on the UI server- and client-side, we might get different results.

**Comment by asaunier** [comment]
So the API would return a "slug" attribute in addition to the standard ones?
That sounds interesting.

**Comment by asaunier** [comment]
On the other hand, this slug thing is a pure UI thing. This attribute is probably useless for other consuming apps.

Actually there's probably not a big risk that the JS slugs are different from the python ones and those JS-generated slugs will probably be not available to the search engines, so that's not that important if they are a bit different. The thing will be to avoid that the slug is missing (which will trigger a redirection).

**Comment by tsauerwein** [comment]
> Actually there's probably not a big risk that the JS slugs are different from the python ones and those JS-generated slugs will probably be not available to the search engines

Can you remind me where slugs are created and used on the client-side?

**Comment by asaunier** [comment]
In cards

**Comment by asaunier** [comment]
For the client-side, https://github.com/lucasconstantino/angular-slug seems a good option because it supports non-latin characters as well (yes, in c2corg there are waypoint or route names with russian, greek, chineese, japanese or arabic characters! eg. http://www.camptocamp.org/routes/632364/en/nagisodake--s-route-to-araragi--onsen-place).

This tool is an angular wrapper around the slug.js library which seems to use a unicode library etc.
But that's not very clear what files must be included in our mako template and if there's a minified version.

**Comment by asaunier** [comment]
This PR is ready for reviewing.

What changes:
* document URLs now contain a slug based upon the document title (title + title_prefix for routes). If the slug is missing, a 301 redirection is performed towards the URL with the slug.
* slugs are added using a python library when URLs are generated by the UI server-side (SEO links, 301 redirections, archive links).
* for cards (advanced search results and associations) and simple search suggestions, the slug has to be generated client-side. A javascript library (+ an angular wrapper) is used to do so. The results for both pyton and JS tools are consistent for latin-letters titles, a bit less for titles with exotic characters (eg. japanese): those characters are converted by the python tool, ignored by the JS tool (even though the latter seems to support some non-latin alphabet such as russian).
* to ease the use of the slug service, I have created an ``app.Url`` service that is used to generate in JS the document URLs with their slugs.
* tests have been adapted (to support the new URL format and the redirection).

Please note that the 301 redirection now uses a standard document request to the API to get the document title and generate its slug. It would be better if it could request a simple API service that returns the minimal info. See https://github.com/c2corg/v6_api/issues/364

File: c2corg_ui/static/js/map/map.js [hunk 1]
```diff
@@ -659,13 +667,17 @@ app.MapController.prototype.handleMapFeatureClick_ = function(event) {
...
   }, this);
   if (feature) {
-    var module = /** @type {string} */(feature.get('module'));
-    var id = feature.get('documentId').toString();
-    var lang = /** @type {string} */(feature.get('lang'));
-    var url = app.utils.buildDocumentUrl(module, id, lang);
-    if (url) {
-      window.location.href = url;
+    var module = feature.get('module');
+    var id = parseInt(feature.get('documentId'), 10);
```
Comment (tsauerwein) [inline c2corg_ui/static/js/map/map.js]: Do you need to use `parseInt`? Isn't it already a number?

File: c2corg_ui/static/js/urlservice.js [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+
+/**
+ * @param {function(string):string} slug angular-slug service.
+ * @constructor
+ * @ngInject
+ * @struct
+ */
+app.Url = function(slug) {
+
+  slug.defaults.mode = 'custom';
```
Comment (tsauerwein) [inline c2corg_ui/static/js/urlservice.js]: Is it correct that you are setting the mode to custom, before defining it?

**Comment by tsauerwein** [comment]
That looks good. Do you want to wait for https://github.com/c2corg/v6_api/issues/364 before merging?

**Comment by asaunier** [comment]
@tsauerwein Thanks for your review. I have made the 2 changes you have pointed.

I would prefer we merge the PR without waiting for https://github.com/c2corg/v6_api/issues/364 because it's not a blocking issue (especially with the caching system) and it's better to go forward, no having this PR stalling until we have time to add this new API service. After merging the PR I will create a UI issue to remind us of this needed optimization.

**Pull request merged**